src/EFCore.Relational/Query/PathSegment.cs

```diff
@@ -7,44 +7,69 @@ namespace Microsoft.EntityFrameworkCore.Query;
 
 /// <summary>
 ///     <para>
-///         A class representing a component of JSON path used in <see cref="JsonQueryExpression" /> or <see cref="JsonScalarExpression" />.
+///         A struct representing a component of JSON path used in <see cref="JsonQueryExpression" /> or <see cref="JsonScalarExpression" />.
 ///     </para>
 ///     <para>
 ///         This type is typically used by database providers (and other extensions). It is generally
 ///         not used in application code.
 ///     </para>
 /// </summary>
-public class PathSegment
+public struct PathSegment
 {
     /// <summary>
-    ///     Creates a new instance of the <see cref="PathSegment" /> class.
+    ///     Creates a new <see cref="PathSegment" /> struct representing JSON property access.
     /// </summary>
-    /// <param name="key">A key which is being accessed in the JSON.</param>
-    public PathSegment(string key)
+    /// <param name="propertyName">A name of JSON property which is being accessed.</param>
+    public PathSegment(string propertyName)
     {
-        Key = key;
+        PropertyName = propertyName;
+        ArrayIndex = null;
     }
 
     /// <summary>
-    ///     The key which is being accessed in the JSON.
+    ///     Creates a new <see cref="PathSegment" /> struct representing JSON array element access.
     /// </summary>
-    public virtual string Key { get; }
+    /// <param name="arrayIndex"><see langword="abstract"/>An index of an element which is being accessed in the JSON array.</param>
+    public PathSegment(SqlExpression arrayIndex)
+    {
+        ArrayIndex = arrayIndex;
+        PropertyName = null;
+    }
+
+    /// <summary>
+    ///     The name of JSON property which is being accessed.
+    /// </summary>
+    public string? PropertyName { get; }
+
+    /// <summary>
+    ///     The index of an element which is being accessed in the JSON array.
+    /// </summary>
+    public SqlExpression? ArrayIndex { get; }
 
     /// <inheritdoc />
     public override string ToString()
-        => (Key == "$" ? "" : ".") + Key;
+    {
+        var arrayIndex = ArrayIndex switch
+        {
+            null => "",
+            SqlConstantExpression sqlConstant when sqlConstant.Value != null => $"[{sqlConstant.Value}]",
+            SqlParameterExpression sqlParameter => $"[{sqlParameter.Name}]",
+            _ => "[(...)]"
+        };
+
+        return (PropertyName == "$" ? "" : ".") + (PropertyName ?? "") + arrayIndex;
```

Comment:
There's no situation where both PropertyName and ArrayIndex are set, right? If so, then:

```suggestion
        return (PropertyName == "$" ? "" : ".") + (PropertyName ?? arrayIndex);
```